kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelService.kt

```diff
@@ -317,20 +326,40 @@ class TunnelService : VpnService() {
             tunnelState = State.CONNECTING
             updateStatusNotification(TunnelStatusNotification.Connecting)
 
-            connlibSessionPtr =
-                ConnlibSession.connect(
-                    apiUrl = config.apiUrl,
-                    token = token,
-                    deviceId = deviceId(),
-                    deviceName = getDeviceName(),
-                    osVersion = Build.VERSION.RELEASE,
-                    logDir = getLogDir(),
-                    logFilter = config.logFilter,
-                    callback = callback,
-                )
-
-            startNetworkMonitoring()
-            startDisconnectMonitoring()
+            val executor = Executors.newSingleThreadExecutor()
+
+            executor.execute {
+                val deviceInfo = DeviceInfo()
+
+                runCatching {
+                    Tasks.await(FirebaseInstallations.getInstance().id)
+                }.onSuccess { firebaseInstallationId ->
+                    deviceInfo.firebaseInstallationId = firebaseInstallationId
+                }.onFailure { exception ->
+                    Log.d(TAG, "Failed to obtain firebase installation id: $exception")
+                }
+
+                val gson: Gson =
+                    GsonBuilder()
+                        .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
+                        .create()
+
+                connlibSessionPtr =
+                    ConnlibSession.connect(
+                        apiUrl = config.apiUrl,
+                        token = token,
+                        deviceId = deviceId(),
+                        deviceName = getDeviceName(),
+                        osVersion = Build.VERSION.RELEASE,
+                        logDir = getLogDir(),
+                        logFilter = config.logFilter,
+                        callback = callback,
+                        deviceInfo = gson.toJson(deviceInfo),
+                    )
+
+                startNetworkMonitoring()
+                startDisconnectMonitoring()
+            }
```

Comment:
This has gotten a bit verbose which was the main reason I suggested to do it elsewhere. I don't know how long it takes to fetch the Firebase ID, I guess once it is cached locally it shouldn't delay the connect much?